[WebGPU][New test] resource_compatibility is failing - gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,render_pipeline,resource_compatibility:*

mwyrzykowski opened the pull request:
#### ad4f7e712c10a2288245b1ed2d6faa628237522f
<pre>
[WebGPU][New test] resource_compatibility is failing - gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,render_pipeline,resource_compatibility:*
<a href="https://bugs.webkit.org/show_bug.cgi?id=280784">https://bugs.webkit.org/show_bug.cgi?id=280784</a>
<a href="https://rdar.apple.com/135707855">rdar://135707855</a>

Reviewed by Dan Glastonbury.

These tests were added to the CTS after we made all the validation ones passing.

Add some logic and error messages for resource compatibility between shader texture types
and the corresponding layouts.

* LayoutTests/http/tests/webgpu/webgpu/api/validation/compute_pipeline-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/render_pipeline/resource_compatibility-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:
Add passing expectations.

* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::textureKindEqualsViewDimension):
(WGSL::depthTextureKindEqualsViewDimension):
(WGSL::nameForBindingType):
(WGSL::nameForTextureSampleType):
(WGSL::nameForTextureDepthKind):
(WGSL::nameForTextureKind):
(WGSL::nameForTextureViewDimension):
(WGSL::nameForPrimitiveKind):
(WGSL::errorValidatingTexture):
(WGSL::storageTextureKindEqualsViewDimension):
(WGSL::isStorageTexture):
(WGSL::errorValidatingTypes):
(WGSL::errorValidatingVariableAndEntryMatch):
(WGSL::RewriteGlobalVariables::insertStructs):
(WGSL::isTexture): Deleted.
(WGSL::typesMatch): Deleted.
(WGSL::variableAndEntryMatch): Deleted.
* Source/WebGPU/WebGPU/ComputePipeline.mm:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::wgslViewDimension):
* Source/WebKit/Shared/WebGPU/WebGPUTextureBindingLayout.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):

Canonical link: <a href="https://commits.webkit.org/284818@main">https://commits.webkit.org/284818@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1cdfd3aaa80ea9052600df856225379afbb6716

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70592 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49998 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74681 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21770 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57798 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21610 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55913 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14382 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73658 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45460 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60854 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36374 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42118 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18290 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20131 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64052 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18645 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76401 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14818 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17858 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63640 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14861 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60922 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63577 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15637 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11626 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5271 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45800 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/568 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46874 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48151 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46616 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -1612,14 +1612,278 @@ static bool isSampler(const AST::Variable& variable, SamplerBindingType bindingT
+    }
+
+    if (!result)
+        return makeString("match6 failed, sampleType "_s, nameForTextureSampleType(sampleType));
```
need to clean up this error message

djg commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -1612,14 +1612,278 @@ static bool isSampler(const AST::Variable& variable, SamplerBindingType bindingT
 }
 
-static bool isStorageTexture(const AST::Variable& variable, StorageTextureAccess textureAccess)
+static String nameForBindingType(BindingType bindingType)
```
Should this be AsciiLiteral instead of String?

mwyrzykowski replied:
Thank you, didn't know about AsciiLiteral before this :)

djg commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -1636,37 +1900,38 @@ static bool isExternalTexture(const AST::Variable& variable)
 }
 
-static bool variableAndEntryMatch(const AST::Variable& variable, const BindGroupLayoutEntry& entry)
+static std::optional<String> errorValidatingVariableAndEntryMatch(const AST::Variable& variable, const BindGroupLayoutEntry& entry)
```
Does the empty String work for `std::nullopt` case?

mwyrzykowski replied:
Yes, assuming I change line 2022:
```
if (auto error = errorValidatingVariableAndEntryMatch(*variable, *entryIt->value))
```
to something like:
```
if (auto error = errorValidatingVariableAndEntryMatch(*variable, *entryIt->value); error.length())
```

djg commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -1636,37 +1900,38 @@ static bool isExternalTexture(const AST::Variable& variable)
 }
 
-static bool typesMatch(const AST::Variable& variable, const BindGroupLayoutEntry::BindingMember& bindingMember)
+static std::optional<String> errorValidatingTypes(const AST::Variable& variable, const BindGroupLayoutEntry::BindingMember& bindingMember)
```
Does the empty String work for `std::nullopt` case?

mwyrzykowski replied:
oh yes, we could use empty string instead

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU][New test] resource_compatibility is failing - gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,render_pipeline,resource_compatibility:*
https://bugs.webkit.org/show_bug.cgi?id=280784
rdar://135707855

Reviewed by Dan Glastonbury.

These tests were added to the CTS after we made all the validation ones passing.

Add some logic and error messages for resource compatibility between shader texture types
and the corresponding layouts.

* LayoutTests/http/tests/webgpu/webgpu/api/validation/compute_pipeline-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/validation/render_pipeline/resource_compatibility-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:
Add passing expectations.

* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::textureKindEqualsViewDimension):
(WGSL::depthTextureKindEqualsViewDimension):
(WGSL::nameForBindingType):
(WGSL::nameForTextureSampleType):
(WGSL::nameForTextureDepthKind):
(WGSL::nameForTextureKind):
(WGSL::nameForTextureViewDimension):
(WGSL::nameForPrimitiveKind):
(WGSL::errorValidatingTexture):
(WGSL::storageTextureKindEqualsViewDimension):
(WGSL::isStorageTexture):
(WGSL::errorValidatingTypes):
(WGSL::errorValidatingVariableAndEntryMatch):
(WGSL::RewriteGlobalVariables::insertStructs):
(WGSL::isTexture): Deleted.
(WGSL::typesMatch): Deleted.
(WGSL::variableAndEntryMatch): Deleted.
* Source/WebGPU/WebGPU/ComputePipeline.mm:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::wgslViewDimension):
* Source/WebKit/Shared/WebGPU/WebGPUTextureBindingLayout.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):

Canonical link: https://commits.webkit.org/284818@main
